pgjdbc/src/main/java/org/postgresql/sasl/ScramAuthenticator.java

```diff
@@ -0,0 +1,154 @@
+ * See the LICENSE file in the project root for more information.
+ */
+
+
+package org.postgresql.sasl;
+
+import org.postgresql.core.PGStream;
+import org.postgresql.util.GT;
+import org.postgresql.util.PSQLException;
+import org.postgresql.util.PSQLState;
+
+import com.ongres.scram.client.ScramClient;
+import com.ongres.scram.client.ScramSession;
+import com.ongres.scram.common.ScramMechanisms;
+import com.ongres.scram.common.exception.ScramException;
+import com.ongres.scram.common.exception.ScramInvalidServerSignatureException;
+import com.ongres.scram.common.exception.ScramParseException;
+import com.ongres.scram.common.exception.ScramServerErrorException;
+import com.ongres.scram.common.stringprep.StringPreparations;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class ScramAuthenticator {
+  private static final Logger LOGGER = Logger.getLogger(ScramAuthenticator.class.getName());
+  private static final String EXPECTED_SERVER_ADVERTISED_METHODS = "SCRAM-SHA-256";
+
+  private final String user;
+  private final String password;
+  private final PGStream pgStream;
+  private final ScramClient scramClient = ScramClient
+      .channelBinding(ScramClient.ChannelBinding.NO)
+      .stringPreparation(StringPreparations.NO_PREPARATION)
+      .selectClientMechanism(ScramMechanisms.SCRAM_SHA_256)
+      .setup();
+  private ScramSession scramSession;
+  private ScramSession.ServerFirstProcessor serverFirstProcessor;
+  private ScramSession.ClientFinalProcessor clientFinalProcessor;
+
+  private static void log(Level level, String message, Object... args) {
+    if (LOGGER.isLoggable(level)) {
+      LOGGER.log(level, message, args);
+    }
+  }
+
+  private static void logFinest(String message, Object... args) {
+    log(Level.FINEST, message, args);
+  }
+
+  private static void throwConnectionRejectedException(String message, Object... args)
+      throws PSQLException {
+    throw new PSQLException(
+        GT.tr(message, args),
+        PSQLState.CONNECTION_REJECTED
+    );
+  }
+
+  @FunctionalInterface
+  private interface BodySender {
+    void sendBody(PGStream pgStream) throws IOException;
+  }
+
+  private void authenticationMessageSender(BodySender bodySender, int bodyLength)
+      throws IOException {
+    pgStream.sendChar('p');
+    pgStream.sendInteger4(Integer.BYTES + bodyLength);
+    bodySender.sendBody(pgStream);
+    pgStream.flush();
+  }
+
+  public ScramAuthenticator(String user, String password, PGStream pgStream) {
+    this.user = user;
+    this.password = password;
+    this.pgStream = pgStream;
+  }
+
+  public void processServerMechanismsAndInit() throws IOException, PSQLException {
+    String scramAuthenticationMechanism = pgStream.receiveString();
+    pgStream.receiveChar(); // '0' terminating the String
+    if (!EXPECTED_SERVER_ADVERTISED_METHODS.equals(scramAuthenticationMechanism)) {
```

Comment:
Currently the server only supports a single SCRAM auth mechanism but the protocol spec supports a list of values. This code will break if more than one auth mechanism is sent as the second and onward will not be read (and thus mess up whatever is next on the wire).

We'll need to keep reading strings until you get to one that is of length zero (i.e. the \0 all by itself) to indicate the end of the list. The strings themselves should be added to a list so that the priority order can be maintained.